File: tools/editor/js/pickmap.js
// The pick map, in the browser: which world tile the cursor is actually on.
//
// Named pickmap rather than pick because shapes.js already exports a `pick` - hit testing a shape
// the editor drew. Two different questions: that one asks what is selected, this one asks where
// the cursor is.
//
// WHAT THIS REPLACES. The isometric projection has no inverse - a screen pixel names a world tile
// only once a Z is assumed, and assuming the ground plane is wrong by z*4/44 tiles, about 2.7 where
// Britain stands. So the art view could select things the editor had drawn itself (it knew where it
// put them) but could not place, drag or draw anything new, because none of those know a Z. The
// renderer knows: it had the answer in its hand while it was drawing. A pick map is that answer,
// recorded per pixel by the same painter's pass and the same floor filter as the picture - so a
// roof at the All stop picks the roof, and the floor under it picks the floor at Ground.
//
// ONE SIDECAR ANSWERS EVERY ZOOM. Pick maps exist only at the deepest level, because the lookup
// goes through a FACET-GLOBAL CANVAS PIXEL, which is the same number whatever the camera is doing.
// See iso.canvasToTile and tools/MapExport/PickMap.cs.
//
// FETCHED, NOT DRAWN. The bytes come through fetch().arrayBuffer(), gzipped on the wire and
// inflated by the browser, so they are exact. An image would have to come back through a canvas,
// and that path premultiplies alpha and applies colour management - a pick map that is nearly right
// is a waypoint three tiles from where it was clicked.
//
// LAZY, one tile at a time. A pick map is only wanted for tiles somebody puts the cursor on, which
// is a small fraction of what gets drawn; a 256-pixel sidecar covers about twelve world tiles at
// 1:1, so crossing into a new one is a rare event and costs about 13 ms of renderer time once,
// ever. Neighbours are deliberately NOT prefetched: the renderer is serial and its queue is LIFO,
// so eight speculative tiles would sit in front of the one the cursor is actually on.

import * as iso from './iso.js';

/** "GGPK" - the sidecar's magic, checked so a wrong file is a refusal rather than a wrong answer. */
const MAGIC = [0x47, 0x47, 0x50, 0x4B];

/** Bytes before the first plane. Must match PickMap.HeaderSize. */
const HEADER = 16;

export const KIND_NONE = 0;
export const KIND_LAND = 1;
export const KIND_STATIC = 2;
export const KIND_ITEM = 3;

const KIND_NAMES = { 1: 'land', 2: 'static', 3: 'item' };

/** key -> { state: 'loading' | 'ready' | 'error', map, reason } */
const cache = new Map();

/**
 * The last thing that went wrong, or null.
 *
 * Kept rather than thrown because the caller is a mousemove: a pick that cannot be had has to
 * degrade to the ground-plane estimate and say so once, not raise on every frame.
 */
let lastProblem = null;

export function problem() {
    return lastProblem;
}

/**
 * Forgets everything. A new world-item snapshot is a whole new set of URLs, so the entries under
 * the old one are dead weight rather than wrong - but there can be tens of thousands of them over a
 * long session, and they are 256 KB each.
 */
export function reset() {
    cache.clear();
    lastProblem = null;
}

export function cached() {
    return cache.size;
}

/**
 * Where a pick map for a canvas pixel lives, or null when the view cannot address one.
 *
 * The floor is the EFFECTIVE floor - what is actually drawn - rather than the slider's position, so
 * the pick agrees with the picture at zooms where the slider has gone inactive.
 */
function locate(view, canvasX, canvasY) {
    if (!view.art || !view.facet) {
        return null;
    }

    const level = view.art.pickLevel;

    if (!Number.isInteger(level)) {
        return null;
    }

    const at = iso.canvasToTile(
        canvasX, canvasY, level, view.artMaxLevel, view.art.tileSize || iso.TILE_SIZE);

    if (at.tileX < 0 || at.tileY < 0) {
        return null;
    }

    const layer = view.items ? `items-${view.items.id}` : 'map';
    const floor = view.effectiveFloor;
    const key = `${view.facet.name}/v${view.art.version}/${layer}/${floor}/${level}/${at.tileX}/${at.tileY}`;

    return { key, url: `/tiles/iso/${key}.pick`, pixelX: at.pixelX, pixelY: at.pixelY };
}

/**
 * What the renderer recorded at a canvas pixel, or null.
 *
 * Null means three different things and the caller treats them alike - the tile is still in flight,
 * the tile could not be had, or nothing was drawn at that pixel (open sea, or off the facet). All
 * three mean "no exact answer", and the readout falls back to the estimate and marks it.
 */
export function at(view, canvasX, canvasY) {
    const where = locate(view, canvasX, canvasY);

    if (!where) {
        return null;
    }

    const entry = cache.get(where.key);

    if (!entry || entry.state !== 'ready') {
        return null;
    }

    return entry.map.at(where.pixelX, where.pixelY);
}

/**
 * Asks for the pick map covering a canvas pixel, if it is not already asked for.
 *
 * Fire and forget: `onLoaded` is called once, when a fetch this call started has arrived, so the
 * readout can correct itself under a cursor that has stopped moving.
 */
export function request(view, canvasX, canvasY, onLoaded) {
    const where = locate(view, canvasX, canvasY);

    if (!where || cache.has(where.key)) {
        return;
    }

    load(view, where).then(
        () => { if (onLoaded) { onLoaded(); } },
        () => { if (onLoaded) { onLoaded(); } });
}

/**
 * The pick at a canvas pixel, waiting for its tile if it has to.
 *
 * What a CLICK uses. A click is rare and a pick map is about 13 ms, so waiting is imperceptible and
 * is the only answer that is never wrong; a mousemove uses `at` instead and never waits. In
 * practice the wait is already over, because the mousemove that positioned the cursor asked for
 * this same tile.
 */
export async function resolve(view, canvasX, canvasY) {
    const where = locate(view, canvasX, canvasY);

    if (!where) {
        return null;
    }

    let entry = cache.get(where.key);

    if (!entry) {
        entry = await load(view, where);
    } else if (entry.state === 'loading') {
        await entry.pending.catch(() => {});
        entry = cache.get(where.key);
    }

    return entry && entry.state === 'ready' ? entry.map.at(where.pixelX, where.pixelY) : null;
}

function load(view, where) {
    const entry = { state: 'loading', map: null, reason: null };

    entry.pending = fetch(where.url)
        .then((response) => {
            if (!response.ok) {
                throw new Error(`HTTP ${response.status}`);
            }

            return response.arrayBuffer();
        })
        .then((buffer) => {
            entry.map = decode(buffer, view.art.pickFormat);
            entry.state = 'ready';
            lastProblem = null;
            return entry;
        })
        .catch((error) => {
            entry.state = 'error';
            entry.reason = error.message;
            lastProblem = error.message;
            return entry;
        });

    cache.set(where.key, entry);

    return entry.pending;
}

/**
 * The bytes into planes, or a throw.
 *
 * The format byte is checked against what the renderer said in its handshake. It is NOT a cache
 * key - the file is gzipped whole, so the byte is not readable without inflating it - and changing
 * the format means bumping TileServer.Version, which orphans the sidecars along with the tiles they
 * sit beside. Checking it here is what turns a stale cache into a sentence somebody can act on
 * rather than picks three hundred tiles away.
 */
export function decode(buffer, expectedFormat) {
    const bytes = new Uint8Array(buffer);

    if (bytes.length < HEADER) {
        throw new Error('The pick map is truncated.');
    }

    for (let i = 0; i < MAGIC.length; i++) {
        if (bytes[i] !== MAGIC[i]) {
            throw new Error('That is not a pick map.');
        }
    }

    const format = bytes[4];

    if (expectedFormat !== undefined && format !== expectedFormat) {
        throw new Error(
            `The pick cache is stale - format ${format}, the renderer writes ${expectedFormat}. `
                + 'Bump TileServer.Version or delete tools/editor/tiles.');
    }

    const size = (bytes[6] << 8) | bytes[7];
    const view = new DataView(buffer);
    const baseX = view.getInt32(8, false);
    const baseY = view.getInt32(12, false);
    const plane = size * size;

    if (bytes.length < HEADER + (plane * 4)) {
        throw new Error('The pick map is truncated.');
    }

    const kind = bytes.subarray(HEADER, HEADER + plane);
    const dx = bytes.subarray(HEADER + plane, HEADER + (plane * 2));
    const dy = bytes.subarray(HEADER + (plane * 2), HEADER + (plane * 3));
    const z = bytes.subarray(HEADER + (plane * 3), HEADER + (plane * 4));

    return {
        size,
        format,
        baseX,
        baseY,

        at(pixelX, pixelY) {
            if (pixelX < 0 || pixelY < 0 || pixelX >= size || pixelY >= size) {
                return null;
            }

            const index = (pixelY * size) + pixelX;

            // Nothing was drawn here - open sea, or off the edge of the facet. Not an error, and
            // not a tile: there is genuinely nothing under the cursor to name.
            if (kind[index] === KIND_NONE) {
                return null;
            }

            return {
                x: baseX + dx[index],
                y: baseY + dy[index],
                // Z is an sbyte, stored +128, so this is exact over the whole range.
                z: z[index] - 128,
                kind: kind[index],
                what: KIND_NAMES[kind[index]] || 'none'
            };
        }
    };
}

File: tools/MapExport/PickMap.cs
using System;
using System.IO;
using System.IO.Compression;

namespace Server.Custom.MapExport
{
    /// <summary>
    /// Which world tile, and at what standing Z, each pixel of a rendered tile belongs to.
    ///
    /// WHY THIS EXISTS. The isometric projection has no inverse. A screen pixel names a world tile
    /// only once a Z is assumed, and assuming the ground plane is wrong by z*4/44 tiles - about 2.7
    /// where Britain stands, near z 30. That is why the editor refused to place or drag anything in
    /// the art view: a waypoint dropped three tiles from where it was clicked would look right and
    /// be wrong. A pick map turns the inverse into a lookup, by recording the answer while the
    /// renderer is drawing and therefore already knows it.
    ///
    /// It is written by the SAME painter's pass and the SAME floor filter as the picture, at the
    /// same two pixel-write sites (IsoTileRenderer.Blit and IsoTileRenderer.Triangle). So a click
    /// lands on the tile the eye sees - the roof at the All stop, and the floor under it at Ground,
    /// because a roof the floor filter dropped was never drawn and so was never recorded.
    ///
    /// ONE RESOLUTION ANSWERS EVERY ZOOM. A pick map is rendered only at the deepest level, 1:1,
    /// and it is read by turning a screen point into a FACET-GLOBAL CANVAS PIXEL, which is the same
    /// number at every zoom. So there is no second copy at 1:2 and no decimation question - which
    /// matters, because world coordinates cannot be box-averaged the way colour can, and any rule
    /// for picking one of four would be a lie about the other three.
    ///
    /// ---
    ///
    /// THE FORMAT, gzip-compressed whole and served with Content-Encoding: gzip, so the browser
    /// decompresses it natively and fetch().arrayBuffer() hands over the exact bytes:
    ///
    ///      0  "GGPK"                 magic
    ///      4  u8   format            Format, below
    ///      5  u8   reserved          0
    ///      6  u16  size              pixels per side
    ///      8  i32  baseX             the world x that dx 0 means
    ///     12  i32  baseY
    ///     16  size*size  kind        0 nothing drawn, 1 land, 2 static, 3 world item
    ///        +size*size  dx          worldX - baseX
    ///        +size*size  dy          worldY - baseY
    ///        +size*size  z           standing Z + 128
    ///
    /// NOT A PNG, for two reasons. Getting EXACT bytes back out of a PNG in a browser means a
    /// canvas round trip, and that path premultiplies alpha and applies colour management - and a
    /// pick map that is nearly right is a waypoint three tiles out. And an RGBA8 pixel cannot hold
    /// what is needed anyway: world x is 13 bits, y is 12 and Z is 8, which is 33.
    ///
    /// PLANAR, NOT INTERLEAVED, because each plane is piecewise constant over a drawn sprite - one
    /// land diamond is one value in all four - and deflate finds those runs only when they are
    /// adjacent. Interleaving shreds every run into stripes of four.
    ///
    /// Z IS EXACT IN A BYTE because Z is an sbyte throughout UO, so +128 covers -128..127 with
    /// nothing left over and nothing to clamp.
    ///
    /// dx AND dy FIT IN A BYTE, and the writer refuses rather than wrapping if they ever do not. A
    /// 256-pixel tile is about twelve world tiles across, but IsoTransform.Footprint widens the
    /// columns it walks by ZLift and MaxSpriteHeight - 512 iso pixels each - so a tall sprite
    /// anchored well outside the tile can still paint into it. The real span works out around
    /// sixty tiles per axis, four times inside the byte; the refusal is there because a wrapped dx
    /// would not look like a failure, it would look like a pick three hundred tiles away.
    /// </summary>
    internal sealed class PickMap
    {
        /// <summary>
        /// The sidecar's own format, independent of TileServer.Version.
        ///
        /// The cache path carries the RENDERER version, which orphans a .pick whenever the art
        /// changes - but this format could change while no pixel does, and a file cached under a
        /// version that is still current would then be stale forever. So the bridge reads this byte
        /// out of a cache hit and re-renders when it is not the one the handshake advertised.
        /// </summary>
        public const byte Format = 1;

        public const byte KindNone = 0;
        public const byte KindLand = 1;
        public const byte KindStatic = 2;
        public const byte KindItem = 3;

        /// <summary>The header, in bytes, before the first plane.</summary>
        public const int HeaderSize = 16;

        private static readonly byte[] Magic = { (byte)'G', (byte)'G', (byte)'P', (byte)'K' };

        private readonly int _size;
        private readonly byte[] _kind;
        private readonly int[] _x;
        private readonly int[] _y;
        private readonly int[] _z;

        public PickMap(int size)
        {
            if (size <= 0 || size > 0xFFFF)
            {
                throw new ArgumentOutOfRangeException("size");
            }

            _size = size;
            _kind = new byte[size * size];
            _x = new int[size * size];
            _y = new int[size * size];
            _z = new int[size * size];
        }

        public int Size { get { return _size; } }

        /// <summary>
        /// Records a pixel. Called from the two blit inner loops, so it does no work a caller could
        /// have hoisted: the cell is constant for a whole sprite and is built once per sprite.
        /// </summary>
        public void Set(int index, PickCell cell)
        {
            _kind[index] = cell.Kind;
            _x[index] = cell.X;
            _y[index] = cell.Y;
            _z[index] = cell.Z;
        }

        /// <summary>What is recorded at a pixel, for --pick and for the tests.</summary>
        public PickCell At(int pixelX, int pixelY)
        {
            int index = (pixelY * _size) + pixelX;

            return new PickCell(_x[index], _y[index], _z[index], _kind[index]);
        }

        public static string KindName(byte kind)
        {
            switch (kind)
            {
                case KindLand: return "land";
                case KindStatic: return "static";
                case KindItem: return "item";
                default: return "none";
            }
        }

        /// <summary>
        /// Writes the file, creating the directory. Returns the compressed size on disk.
        ///
        /// The whole file is gzipped rather than each plane, so the planes are adjacent in the
        /// stream and a run that spans two of them costs nothing extra.
        /// </summary>
        public long Write(string path)
        {
            int baseX, baseY;
            Bounds(out baseX, out baseY);

            var raw = new byte[HeaderSize + (_kind.Length * 4)];

            Array.Copy(Magic, 0, raw, 0, Magic.Length);
            raw[4] = Format;
            raw[5] = 0;
            raw[6] = (byte)(_size >> 8);
            raw[7] = (byte)_size;
            WriteInt32(raw, 8, baseX);
            WriteInt32(raw, 12, baseY);

            int kindAt = HeaderSize;
            int dxAt = kindAt + _kind.Length;
            int dyAt = dxAt + _kind.Length;
            int zAt = dyAt + _kind.Length;

            for (int i = 0; i < _kind.Length; i++)
            {
                if (_kind[i] == KindNone)
                {
                    // Nothing drawn here. The three coordinate planes stay 0 rather than carrying
                    // a sentinel: `kind` is the sentinel, and zeroes compress to nothing.
                    continue;
                }

                int dx = _x[i] - baseX;
                int dy = _y[i] - baseY;

                if (dx < 0 || dx > 255 || dy < 0 || dy > 255)
                {
                    throw new InvalidOperationException(String.Format(
                        "A pick tile spans more than 256 world tiles ({0},{1} against base {2},{3}). "
                            + "See the note on dx/dy in PickMap.",
                        _x[i], _y[i], baseX, baseY));
                }

                raw[kindAt + i] = _kind[i];
                raw[dxAt + i] = (byte)dx;
                raw[dyAt + i] = (byte)dy;
                raw[zAt + i] = (byte)(_z[i] + 128);
            }

            string directory = Path.GetDirectoryName(path);

            if (!String.IsNullOrEmpty(directory))
            {
                Directory.CreateDirectory(directory);
            }

            using (var file = new FileStream(path, FileMode.Create, FileAccess.Write, FileShare.None))
            {
                using (var gzip = new GZipStream(file, CompressionMode.Compress, true))
                {
                    gzip.Write(raw, 0, raw.Length);
                }
            }

            return new FileInfo(path).Length;
        }

        /// <summary>
        /// The lowest world x and y anything was recorded at. An empty map bases at 0,0 - nothing
        /// reads a plane whose kind is None, so the base is arbitrary there.
        /// </summary>
        private void Bounds(out int baseX, out int baseY)
        {
            baseX = Int32.MaxValue;
            baseY = Int32.MaxValue;

            for (int i = 0; i < _kind.Length; i++)
            {
                if (_kind[i] == KindNone)
                {
                    continue;
                }

                if (_x[i] < baseX) { baseX = _x[i]; }
                if (_y[i] < baseY) { baseY = _y[i]; }
            }

            if (baseX == Int32.MaxValue)
            {
                baseX = 0;
                baseY = 0;
            }
        }

        private static void WriteInt32(byte[] buffer, int offset, int value)
        {
            buffer[offset] = (byte)(value >> 24);
            buffer[offset + 1] = (byte)(value >> 16);
            buffer[offset + 2] = (byte)(value >> 8);
            buffer[offset + 3] = (byte)value;
        }
    }

    /// <summary>
    /// One pixel's answer. A struct passed by value because it is built once per SPRITE and read
    /// once per pixel of it - a sprite covers one world column at one Z, so every pixel it owns
    /// carries the same three numbers.
    /// </summary>
    internal struct PickCell
    {
        public readonly int X;
        public readonly int Y;
        public readonly int Z;
        public readonly byte Kind;

        public PickCell(int x, int y, int z, byte kind)
        {
            X = x;
            Y = y;
            Z = z;
            Kind = kind;
        }
    }
}
